sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -0,0 +1,1512 @@
+
+# Those constants have been chosen for modern laptops' caches and architecture.
+DEF CHUNK_SIZE = 256  # number of vectors
+DEF MIN_CHUNK_SAMPLES = 20
+
+
+# TODO: change for `libcpp.algorithm.move` once Cython 3 is used
+# Introduction in Cython:
+# https://github.com/cython/cython/blob/05059e2a9b89bf6738a7750b905057e5b1e3fe2e/Cython/Includes/libcpp/algorithm.pxd#L47 #noqa
+cdef extern from "<algorithm>" namespace "std" nogil:
+    OutputIt move[InputIt, OutputIt](InputIt first, InputIt last, OutputIt d_first) except + #noqa
+
+######################
+## std::vector to np.ndarray coercion
+# As type covariance is not supported for C++ containers via Cython,
+# we need to redefine fused types.
+ctypedef fused vector_DITYPE_t:
+    vector[ITYPE_t]
+    vector[DTYPE_t]
+
+
+ctypedef fused vector_vector_DITYPE_t:
+    vector[vector[ITYPE_t]]
+    vector[vector[DTYPE_t]]
+
+
+cdef extern from "numpy/arrayobject.h":
+    int PyArray_SetBaseObject(np.ndarray arr, PyObject *obj) nogil except -1
+
+
+cdef class StdVectorSentinel:
+    """Wraps a reference to a vector which will be deallocated with this object.
+
+    When created, the StdVectorSentinel swaps the reference of its internal
+    vectors with the provided one (vec_ptr), thus making the StdVectorSentinel
+    manage the provided one's lifetime.
+    """
+    pass
+
+# We necessarily need to define two extension types extending StdVectorSentinel
+# because we need to provide the dtype of the vector but can't use numeric fused types.
+cdef class StdVectorSentinelDTYPE(StdVectorSentinel):
+    cdef vector[DTYPE_t] vec
+
+    @staticmethod
+    cdef StdVectorSentinel create_for(vector[DTYPE_t] * vec_ptr):
+        # This initializes the object directly without calling __init__
+        cdef StdVectorSentinelDTYPE sentinel = StdVectorSentinelDTYPE.__new__(StdVectorSentinelDTYPE)
+        sentinel.vec.swap(deref(vec_ptr))
+        return sentinel
+
+
+cdef class StdVectorSentinelITYPE(StdVectorSentinel):
+    cdef vector[ITYPE_t] vec
+
+    @staticmethod
+    cdef StdVectorSentinel create_for(vector[ITYPE_t] * vec_ptr):
+        # This initializes the object directly without calling __init__
+        cdef StdVectorSentinelITYPE sentinel = StdVectorSentinelITYPE.__new__(StdVectorSentinelITYPE)
+        sentinel.vec.swap(deref(vec_ptr))
+        return sentinel
+
+
+cdef np.ndarray vector_to_nd_array(vector_DITYPE_t * vect_ptr):
+    """Create a numpy ndarray given a C++ vector.
+
+    The numpy array buffer is the one of the C++ vector.
+    A StdVectorSentinel is registers as the base object for the numpy array,
+    freeing the C++ vector it encapsulates when the numpy array is freed.
+    """
+    typenum = DTYPECODE if vector_DITYPE_t is vector[DTYPE_t] else ITYPECODE
+    cdef:
+        np.npy_intp size = deref(vect_ptr).size()
+        # TODO: use PyArray_SimpleNewFromData from the Numpy C API directly
+        # I've tried, but Cython fails when parsing the C API
```

Comment:
Maybe you could report a simple reproducer to Cython developers and link to the issue from this comment to make it easier to know when this issue could be tackled in the future?

Reply:
IIRC, I noted this `TODO` thinking that using the C API would allow using a `nogil`-context, but we _do need_ the GIL for `Py_INCREF` in any case. Also, we better use `PyArray_SetBaseObject` from the Cython API (via `np.`) as it's currently the case for `PyArray_SimpleNewFromData`, done in 2dcac3f518fc82c05daa34033a6918641d41c7d0.

Thus, I don't think reporting this make sense anymore. WDYT?